cmd/srcd/cmd/sql_test.go

```diff
@@ -5,24 +5,225 @@ package cmd
+	// test that the code that processes each read line is working as expected.
+
+	require := s.Require()
+
+	var out, in bytes.Buffer
+
+	command := s.CommandContext(context.TODO(), "sql")
+	command.Stdout = &out
+	command.Stderr = &out
+	command.Stdin = &in
+
+	io.WriteString(&in, "show tables;\n")
+	io.WriteString(&in, "describe table repositories;\n")
+
+	err := command.Start()
+	require.NoError(err)
+
+	err = command.Wait()
+	require.NoError(err)
+
+	require.Contains(out.String(), showTablesOutput)
+	require.Contains(out.String(), showRepoTableDescOutput)
+}
+
+func (s *SQLREPLTestSuite) TestInteractiveREPL() {
+	require := s.Require()
+
+	command, stdin, ch := s.runInteractiveRepl()
+
+	res := s.runInteractiveQuery(stdin, "show tables;\n", ch)
+	require.Contains(res, showTablesOutput)
+
+	res = s.runInteractiveQuery(stdin, "describe table repositories;\n", ch)
+	require.Contains(res, showRepoTableDescOutput)
+
+	require.NoError(s.exitInteractiveAndWait(10*time.Second, stdin, ch))
+	require.NoError(s.waitMysqlCliContainerStopped(10 * time.Second))
+
+	command.Wait()
+}
+
+func (s *SQLREPLTestSuite) runInteractiveRepl() (*exec.Cmd, *os.File, <-chan string) {
+	s.T().Helper()
+
+	command := s.CommandContext(context.TODO(), "sql")
+
+	ch := make(chan string)
+	cr := cmdtest.NewChannelWriter(ch)
+
+	command.Stdout = cr
+	command.Stderr = cr
+
+	f, err := pty.Start(command)
+	if err != nil {
+		panic(err)
+	}
+
+	mysqlCliReady := make(chan struct{})
+
+	go func() {
+		for s := range ch {
+			lines := strings.Split(s, "\n")
+			for _, l := range lines {
+				if strings.HasPrefix(l, "mysql>") {
+					mysqlCliReady <- struct{}{}
+					return
+				}
+			}
+		}
+	}()
+
+	<-mysqlCliReady
+
+	return command, f, ch
+}
+
+func (s *SQLREPLTestSuite) runInteractiveQuery(stdin *os.File, query string, ch <-chan string) string {
+	io.WriteString(stdin, query)
+
+	var res strings.Builder
+	for c := range ch {
+		fmt.Printf("Message: %s\n", c)
+		res.WriteString(c)
+		if s.containsSQLOutput(res.String()) {
+			break
+		}
+	}
+
+	return res.String()
+}
+
+func (s *SQLREPLTestSuite) exitInteractiveAndWait(timeout time.Duration, stdin *os.File, ch <-chan string) error {
+	io.WriteString(stdin, "exit;\n")
+
+	done := make(chan struct{})
+	go func() {
+		for c := range ch {
+			if strings.Contains(c, "Bye") {
+				done <- struct{}{}
+				// don't return in order to consume all output and let the process exit
+			}
+		}
+	}()
+
+	select {
+	case <-done:
+		stdin.Close()
+		return nil
+	case <-time.After(timeout):
+		return fmt.Errorf("timeout of %v elapsed while waiting to exit", timeout)
+	}
+}
+
+func (s *SQLREPLTestSuite) waitMysqlCliContainerStopped(timeout time.Duration) error {
+	done := make(chan struct{})
+	go func() {
+		for {
+			running, err := docker.IsRunning(components.MysqlCli.Name, "")
+			if !running {
+				done <- struct{}{}
+				return
+			}
+
+			if err != nil {
+				panic(err)
+			}
+
+			time.Sleep(timeout / 5)
```

Comment:
it's not very correct. Example:
timeout = 10s
1-8s - exiting
9s - exited

but this code will never know about it and fail with error. `time.After(timeout)` is going to fire much earlier than `docker.IsRunning` would return results.

Reply:
The idea here is to try to check `docker.IsRunning` a couple of times within the `timeout`. I don't know if I understood what you mean.

Reply:
from the signature `waitMysqlCliContainerStopped(timeout time.Duration)` I understand that function will return `nil` as long as mysqlCli stopped during `timeout`. Which is not true according to the code. It will return `nil` only during `4*timeout/5 + few ms`. Tbh I feel it is a nitpick so you can just ignore it. But in case you didn't notice this problem during implementation I point to it here.

Reply:
Oh ok, I understand what you mean, yeah here the number `5` is just a random number to test multiple times. BTW I simplified it and now it's much more straightforward and clear. Fixed [here](https://github.com/src-d/engine/pull/398/commits/bb507a56626322b77b07452834d641894201979c).